### ISSUE TITLE
git_remote_list does not detect config changes on disk

### ISSUE THREAD

**Issue by Mr-Wallet** [libgit2/libgit2#5167]
### Reproduction steps
1. `git_repository_open` a repository twice (`A` and `B`).
2. `git_remote_create` on repo `A`.
3. `git_remote_list` on repo `B`.

### Expected behavior
The created remote should be listed in both repositories.

### Actual behavior
Only repo `A` sees the new repository.
Repo `B` will not list the new repository.

I have a [branch on my fork](https://github.com/Mr-Wallet/libgit2/tree/issue/multi-thread-remote-list) which includes a [test](https://github.com/Mr-Wallet/libgit2/commit/7ee3eed88fab61017b71455197cdf9c0a514884a) reproducing the issue.

Repo `B` _can_ be induced to bust its config cache with `git_repository_config_snapshot`, but this is burdensome if I'm not interested in the config, because I need to immediately free it.

This issue is a concern when multi-threading, where object shouldn't be shared between threads. After I finish creating a remote on one thread, I would expect to be able to read the remote list on another thread at any point, and get accurate results.

I'm not very familiar with LibGit2, so maybe there's already a cache-busting mechanism, but I was unable to find any mention in the API.

Assuming this isn't just a matter of me doing it wrong, some possible solutions:
- Always bust the cache on `git_remote_list`.
- Always `stat` the config(s) and bust the cache if they've been modified since the last time the configs were parsed.
- Provide an option to `git_remote_list` for busting the cache.
- Directly expose/document a function that busts the cache in the same way that `git_repository_config_snapshot` does, but without taking an `out` parameter which I need to manually free.

### Version of libgit2 (release number or SHA1)
f92d495d44bff2bdb6fd99524a93986ea268c9e8 (`master` as of `2019-07-15T22:00:00Z`)

### Operating system(s) tested
The hyperlinked test was run on Windows 10 x64, but our downstream project ([nodegit](https://github.com/nodegit/nodegit)) showed this behavior on Linux and MacOS as well.

**Comment by pks-t** [comment]
This has recently changed in #5132.

* Always bust the cache on git_remote_list.

We used to do that, as we always hashed the complete config file's contents. We got complaints that this is kind of excessive, and I agree. So we're now doing:

* Always stat the config(s) and bust the cache if they've been modified since the last time the configs were parsed.

We're doing exactly that now for config files. We have a per-config-file stat cache now that verifies that mtime, device and size still match what's expected. If it's all the same, we avoid re-hashing the file contents.

There's some cases though where this is fails: iff using a system with a timestamp resolution of 1s and iff you're rewriting the configuration file to have the exact same size as before, then we will not notice that it's changed. Windows is the most common one where one may not have sub-second timestamp resolution.

* Provide an option to git_remote_list for busting the cache.
* Directly expose/document a function that busts the cache in the same way that git_repository_config_snapshot does, but without taking an out parameter which I need to manually free.

Hum. I feel somehow sympathetic to this request, something like `git_config_invalidate` that busts the config stat cache might be useful. But should the application developer really be expected to call such functions in the few scenarios where it's required?

**Comment by Mr-Wallet** [comment]
Is that "exact same size" measured at KB resolution? Or perhaps the testing environment doesn't actually modify a config on disk? If it were checking the size in bytes, I would expect my example test, which strictly adds a remote, to have busted the cache and succeeded.

**Comment by pks-t** [comment]
> Is that "exact same size" measured at KB resolution? Or perhaps the testing environment doesn't actually modify a config on disk? If it were checking the size in bytes, I would expect my example test, which strictly adds a remote, to have busted the cache and succeeded.

It is in bytes. We do an `fstat` (`GetFileAttributesW` on Windows) on the file and then use the `st_size`, `st_ino` and `st_mtime` fields to determine whether we need to re-read. But it's still quite likely to go wrong in test scenarios: We had some tests where we performed the equivalent of `git_config_set("key", "A")` and `git_config_set("key", "B")` via two different repo handles. It's quite obvious that `st_mtime` and `st_ino` will likely be the same if the filesystem doesn't have sub-second resolution. But in fact, as we changed "key" from "A" to "B", which both have the same length, the size will also be the same.

Long story short: I've cheated a bit in our tests and instead change the key to differently sized strings, which is guaranteed to bust the cache.

> I am amazed that I brought up an issue and most of the fix was already merged 4 days prior!

Heh. I thought it was the other way round: I've implemented the stat cache and four days later we have the first issue reports that it's not working as expected ;) So thanks for the kudos

**Comment by Mr-Wallet** [comment]
The [test](https://github.com/Mr-Wallet/libgit2/commit/7ee3eed88fab61017b71455197cdf9c0a514884a) I've written to reproduce the issue adds a new remote, without modifying anything else. That _should_ make the config strictly larger, shouldn't it? Perhaps the cache is not busting as it was designed to?

**Comment by pks-t** [comment]
> The test I've written to reproduce the issue adds a new remote, without modifying anything else. That should make the config strictly larger, shouldn't it? Perhaps the cache is not busting as it was designed to?

Thanks for the test, I've included it in #5181. The fault here probably wasn't created by #5132 but was a preexisting one. We simply didn't try to refresh config files at all when calling `git_config_iterator`, and that's why we had outdated keys in there. Anyway, this is being fixed by #5181, so I'd appreciate if you could give that one a go and verify that I'm fixing your issue.
